virtual_rainforest/models/soil/carbon.py

```diff
@@ -156,6 +146,73 @@ def mineral_association(
         return temp_scalar * moist_scalar * self.lmwc * (equib_maom - self.maom) / Q_max
 
 
+def calculate_max_sorption_capacity(
+    bulk_density: NDArray[np.float32], percent_clay: NDArray[np.float32]
+) -> NDArray[np.float32]:
+    """Calculate maximum sorption capacity based on bulk density and clay content.
+
+    The maximum sorption capacity is the maximum amount of mineral associated organic
+    matter that can exist per unit volume. This expression and its parameters are also
+    drawn from Mayes et al. (2012). In that paper max sorption also depends on Fe
+    content, but we are ignoring this for now.
+
+    Args:
+        bulk_density: bulk density values for each soil grid cell (kg m^-3)
+        percent_clay: Percentage clay for each soil grid cell
+
+    Returns:
+        Q_max: Maximum sorption capacities (kg m^-3)
+    """
+
+    if np.any(percent_clay > 100.0) or np.any(percent_clay < 0.0):
+        log_and_raise(
+            "Relative clay content must be expressed as a percentage!", ValueError
+        )
+
+    Q_max = bulk_density * 10 ** (
+        MAX_SORPTION_WITH_CLAY["slope"] * np.log10(percent_clay)
+        + MAX_SORPTION_WITH_CLAY["intercept"]
+    )
+    return Q_max
+
+
+def calculate_equilibrium_maom(
+    pH: NDArray[np.float32],
+    Q_max: NDArray[np.float32],
+    lmwc: NDArray[np.float32],
+) -> NDArray[np.float32]:
+    """Based on Langmuir coefficients calculate equilibrium MAOM concentration.
+
+    Args:
+        pH: pH values for each soil grid cell
+        Q_max: Maximum sorption capacities (kg m^-3)
+        lmwc: Low molecular weight carbon pool (kg C m^-3)
+
+    Returns:
+        equib_maom: Equilibrium concentration of Mineral associated organic carbon
+            (MAOM), assuming fixed LMWC (kg C m^-3)
+    """
+
+    binding_coefficient = calculate_binding_coefficient(pH)
+    return (binding_coefficient * Q_max * lmwc) / (1 + lmwc * binding_coefficient)
+
+
+def calculate_binding_coefficient(pH: NDArray[np.float32]) -> NDArray[np.float32]:
+    """Calculate Langmuir binding coefficient based on pH.
+
+    This specific expression and its parameters are drawn from (Mayes et al. (2012)).
+
+    Args:
+        pH: pH values for each soil grid cell
+
+    Returns:
+        binding_coefficient: Langmuir binding coefficients for mineral association of
+            labile carbon (m^3 kg^-1)
+    """
+
+    return 10.0 ** (BINDING_WITH_PH["slope"] * pH + BINDING_WITH_PH["intercept"])
```

Comment:
Ahh right is that the case for everywhere that I'm using dictionaries of constants? Or is this specific to `BINDING_WITH_PH`?

Reply:
And the class would be defined like this

```python
@dataclass
class BindingWithPH:
    """From linear regression (Mayes et al. (2012))."""
    intercept: float = -0.186
    slope: float = -0.216
```

and accessed like this?

```python
return 10.0 ** (BindingWithPH.slope * pH + BindingWithPH.intercept)
```

Reply:
If it is just the `object.thing` notation then there are other options for having `dot` style dictionaries (`dotmap` for example). But @alexdewar did you mean that `BindingWithPH` would also provide `BindingWithPH.calc_binding`?

``` python
@dataclass
class BindingWithPH:
    """From linear regression (Mayes et al. (2012))."""
    intercept: float = -0.186
    slope: float = -0.216
    
    def calc_binding(ph: NDArray) -> NDArray
        return 10.0 ** (self.slope * pH + self.intercept)
```

Reply:
More generally, I think there's an argument for the `constants` module providing something that can be easily serialised to JSON or YAML so that the `constants` can become part of the configuration if needed. `core.constants` might provide a default set, but people might want to play around with the settings.

I've used `dataclasses` for this kind of role here: https://github.com/davidorme/pyrealm/blob/master/pyrealm/param_classes.py

I will accept that the size of the dataclasses may not be ideal 😬

Reply:
@jacobcook1995 Yes, that's essentially what I had in mind, but I'm open to suggestions if something like `dotmap` (which I hadn't heard of till now) is better.

The reason I suggested a dataclass was so you can get better type linting (i.e. `mypy` will be able to see what fields your `BindingWithPH` class has and what types they are). You could also just have them as separate variables.